File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;
import java.io.*;

class Solution {
    static ArrayList<Integer>[] list;
    public int solution(int n, int[][] edge) {
        list = new ArrayList[n+1];
        int answer = 0;
        boolean[] visited = new boolean[n+1];
        int[] distance = new int[n+1];
        
        for(int i=0; i<=n; i++){
            list[i] = new ArrayList<>();
        }
        
        for(int i=0; i<edge.length; i++){
            list[edge[i][0]].add(edge[i][1]);
            list[edge[i][1]].add(edge[i][0]);
        }
        
        
        Queue<Integer> q = new LinkedList<>();
        q.add(1);
        visited[1] = true;
        
        while(!q.isEmpty()){
            int now = q.poll();
            
            for(int i : list[now]){
                if(!visited[i]){
                    visited[i] = true;
                    distance[i] = distance[now] + 1;
                    q.add(i);
                }
            }
        }
        
        //가장 먼 거리 찾기
        int maxDist = -1;
        for(int i : distance){
            maxDist = Math.max(maxDist, i);
        }
        
        //먼 거리에 해당하는 노드 개수 세기
        for(int i=0; i<=n; i++){
            if(distance[i] == maxDist)
                answer++;
        }
        return answer;
    }
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
import heapq

def solution(n, edge):
    answer = 0
    
    # 양방향 가중치 없는 그래프
    graph = [[] for _ in range(n+1)]
    for a, b in edge:
        graph[a].append(b)
        graph[b].append(a)
        
    # 다익스트라
    dist = [float('inf')] * (n+1)
    dist[1] = 0
    heap = [(0,1)] # (거리, 노드)
    
    while heap:
        distance, node = heapq.heappop(heap)
    	
        for next_node in graph[node]:
            new_d = distance + 1 # 가중치 없으니까 1씩 증가
            if new_d < dist[next_node]:
                dist[next_node] = new_d
                heapq.heappush(heap, (new_d, next_node))
        
    max_dist = max(dist[1:])
    answer = sum(1 for d in dist[1:] if d == max_dist)
    return answer